cpmpy/solvers/choco.py

```diff
@@ -0,0 +1,582 @@
+                    return self.chc_model.pow(self.solver_vars(lhs.args[0]), self.solver_vars(lhs.args[1]),
+                                              chcrhs)
+            raise NotImplementedError(
+                "Not a known supported Choco left-hand-side '{}' {}".format(lhs.name, cpm_expr))
+
+        # base (Boolean) global constraints
+        elif isinstance(cpm_expr, GlobalConstraint):
+
+            if cpm_expr.name == 'alldifferent':
+                vars = self.solver_vars(cpm_expr.args)
+                for i in range(len(vars)):
+                    if isinstance(vars[i], int):
+                        vars[i] = self.chc_model.intvar(vars[i], vars[i])  # convert to "variable"
+                    elif not isinstance(vars[i], IntVar):
+                        raise Exception(f"Choco cannot accept alldifferent with: {vars[i]}")
+                return self.chc_model.all_different(vars)
+            elif cpm_expr.name == 'alldifferent_except0':
+                vars = self.solver_vars(cpm_expr.args)
+                for i in range(len(vars)):
+                    if isinstance(vars[i], int):
+                        vars[i] = self.chc_model.intvar(vars[i], vars[i])  # convert to "variable"
+                    elif not isinstance(vars[i], IntVar):
+                        raise Exception(f"Choco cannot accept alldifferent_except0 with: {vars[i]}")
+                return self.chc_model.all_different_except_0(vars)
+            elif cpm_expr.name == 'allequal':
+                vars = self.solver_vars(cpm_expr.args)
+                for i in range(len(vars)):
+                    if isinstance(vars[i], int):
+                        vars[i] = self.chc_model.intvar(vars[i], vars[i])  # convert to "variable"
+                    elif not isinstance(vars[i], IntVar):
+                        raise Exception(f"Choco cannot accept allequal with: {vars[i]}")
+                return self.chc_model.all_equal(vars)
+            elif cpm_expr.name == 'table':
+                assert (len(cpm_expr.args) == 2)  # args = [array, table]
+                array, table = self.solver_vars(cpm_expr.args)
+                return self.chc_model.table(array, table)
+            elif cpm_expr.name == 'InDomain':
+                assert (len(cpm_expr.args) == 2)  # args = [array, table]
+                expr, table = self.solver_vars(cpm_expr.args)
+                return self.chc_model.member(expr, table)
+            elif cpm_expr.name == "cumulative":
+                start, dur, end, demand, cap = self.solver_vars(cpm_expr.args)
+                # Everything given to cumulative in Choco needs to be a variable.
+                # Convert demands to variables
+                if is_num(demand):  # Create list for demand per task
+                    demand = [demand] * len(start)
+                if isinstance(demand, _NumVarImpl):
+                    demand = self.solver_vars(demand)
+                else:
```

Comment:
Demand in the cumulative can be either a list of demands for each task or be static and equal for all of the tasks.
Do you cover the list of NumVars case here?

Reply:
nope, will include